toothpick-kotlin/src/main/java/toothpick/kotlin/ToothpickInjectDelegate.kt

```diff
@@ -0,0 +1,45 @@
+package toothpick.kotlin
+
+import toothpick.Scope
+import kotlin.properties.ReadOnlyProperty
+import kotlin.reflect.KClass
+import kotlin.reflect.KProperty
+
+typealias InputConfigBlock = (InjectConfig.() -> Unit)
+
+open class ToothpickInjectDelegate<OWNER, T : Any>(
+    protected val clazz: KClass<T>,
+    protected val scope: Scope?,
+    protected val name: String? = null,
+    protected val annotationName: KClass<out Annotation>? = null,
+    protected val injectConfigBlock: InputConfigBlock? = null
+) : ReadOnlyProperty<OWNER, T> {
+    private var value: T? = null
+
+    override fun getValue(thisRef: OWNER, property: KProperty<*>): T {
+        return value ?: injectValue(thisRef, property)
+    }
+
+    protected open fun injectValue(thisRef: OWNER, property: KProperty<*>): T {
+        val injectConfig = InjectConfig()
+        if (injectConfigBlock != null) {
+            injectConfig.apply(injectConfigBlock)
+        }
+        val scope = injectConfig.scope ?: scope ?: getScope(thisRef, property) ?: throw IllegalStateException("No valid scope available when attempting to inject $property in $thisRef")
```

Comment:
Let's finish with a `.` and explain how to fix the issue. If there is more than one alternative, let's explain them all.